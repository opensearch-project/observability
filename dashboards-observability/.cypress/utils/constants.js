/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const delay = 1500;
export const COMMAND_TIMEOUT_LONG = 10000;

// trace analytics
export const TRACE_ID = '8832ed6abbb2a83516461960c89af49d';
export const SPAN_ID = 'a673bc074b438374';
export const SERVICE_NAME = 'frontend-client';
export const SERVICE_SPAN_ID = '7df5609a6d104736';

export const testDataSet = [
  {
    mapping_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-service-map-mappings.json',
    data_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-service-map.json',
    index: 'otel-v1-apm-service-map',
  },
  {
    mapping_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000001-mappings.json',
    data_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000001.json',
    index: 'otel-v1-apm-span-000001',
  },
  {
    mapping_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000001-mappings.json',
    data_url: 'https://raw.githubusercontent.com/opensearch-project/observability/main/dashboards-observability/.cypress/utils/otel-v1-apm-span-000002.json',
    index: 'otel-v1-apm-span-000002',
  },
]

export const setTimeFilter = (setEndTime = false, refresh = true) => {
  const startTime = 'Mar 25, 2021 @ 10:00:00.000';
  const endTime = 'Mar 25, 2021 @ 11:00:00.000';
  cy.get('button.euiButtonEmpty[aria-label="Date quick select"]').click();
  cy.get('.euiQuickSelect__applyButton').click();
  cy.get('.euiSuperDatePicker__prettyFormatLink').click();
  cy.get(
    'button.euiDatePopoverButton--start[data-test-subj="superDatePickerstartDatePopoverButton"]'
  ).click();
  cy.get('.euiTab__content').contains('Absolute').click();
  cy.get('input[data-test-subj="superDatePickerAbsoluteDateInput"]')
    .focus()
    .type('{selectall}' + startTime, { force: true });
  if (setEndTime) {
    cy.wait(delay);
    cy.get(
      'button.euiDatePopoverButton--end[data-test-subj="superDatePickerendDatePopoverButton"]'
    ).click();
    cy.wait(delay);
    cy.get('.euiTab__content').contains('Absolute').click();
    cy.get('input[data-test-subj="superDatePickerAbsoluteDateInput"]')
      .focus()
      .type('{selectall}' + endTime, { force: true });
  }
  if (refresh) cy.get('.euiButton__text').contains('Refresh').click();
  cy.wait(delay);
};

// notebooks
export const TEST_NOTEBOOK = 'Test Notebook';
export const SAMPLE_URL = 'https://github.com/opensearch-project/sql/tree/main/sql-jdbc';
export const NOTEBOOK_TEXT = 'Use Notebooks to interactively and collaboratively develop rich reports backed by live data. Common use cases for notebooks includes creating postmortem reports, designing run books, building live infrastructure reports, or even documentation.';
export const OPENSEARCH_URL = 'https://opensearch.org/docs/latest/observability-plugin/notebooks/';
export const MARKDOWN_TEXT = `%md
# Heading 1

#### List and links

* 1
* 2
* [SQL JDBC](${SAMPLE_URL})

---
#### Code block
* Explain SQL
\`\`\`
POST _plugins/_sql/_explain
{
  "query": "SELECT * FROM my-index LIMIT 50"
}
\`\`\`

#### Table
| a1 | b1 | c1 | d1 |
|----|----|----|----|
| a2 | b2 | c2 | d2 |
| a3 | b3 | c3 | d3 |
`

export const SQL_QUERY_TEXT = `%sql
select * from opensearch_dashboards_sample_data_flights limit 20
`

export const PPL_QUERY_TEXT = `%ppl
source=opensearch_dashboards_sample_data_flights
`

export const supressResizeObserverIssue = () => {
  // exception is thrown on loading EuiDataGrid in cypress only, ignore for now
  cy.on('uncaught:exception', (err, runnable) => {
    if (err.message.includes('ResizeObserver loop')) return false;
  });
};

export const verify_traces_spans_data_grid_cols_exists = () => {
  cy.get('.euiDataGridHeaderCell__content').contains('Span ID').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Trace ID').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Operation').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Duration').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Start time').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('End time').should('exist');
  cy.get('.euiDataGridHeaderCell__content').contains('Errors').should('exist');
}

export const count_table_row = (expected_row_count) => {
  cy.get('.euiDataGridHeader [role="columnheader"]').then($el => {
    let colmun_header_count = Cypress.$($el).length;
    let table_grid_cell_count = Cypress.$('[data-test-subj="dataGridRowCell"]').length;
    const total_row_count = table_grid_cell_count / colmun_header_count;
    expect(total_row_count).to.equal(expected_row_count)
  });
}
