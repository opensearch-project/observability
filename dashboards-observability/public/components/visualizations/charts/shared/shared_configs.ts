/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { forIn } from 'lodash';

export const getPlotlySharedConfigs = () => {
  return {
    layout: {
      showlegend: true,
      margin: {
        l: 60,
        r: 30,
        b: 30,
        t: 50,
        pad: 0,
      },
      height: 1180,
      legend: {
        orientation: 'v',
        traceorder: 'normal',
      },
    },
    config: {
      displaylogo: false,
      responsive: true,
    },
  };
};

enum VIS_CATEGORY {
  BASICS = 'Visualizations',
}

export const getPlotlyCategory = (type = 'enum') => {
  switch (type) {
    case 'enum':
      return VIS_CATEGORY;
    case 'list':
      return Object.values(VIS_CATEGORY);
    case 'keys':
      return Object.keys(VIS_CATEGORY);
    default:
      return VIS_CATEGORY;
  }
};
