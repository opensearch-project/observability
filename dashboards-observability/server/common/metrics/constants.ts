/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ComponentType, CounterType } from './types';

export const WINDOW = 3600;
export const INTERVAL = 60;
export const CAPACITY = (WINDOW / INTERVAL) * 2;
export const MILLIS_MULTIPLIER = 1000;

export const COMPONENTS = [
  'application_analytics',
  'operational_panels',
  'event_analytics',
  'notebooks',
  'trace_analytics',
  'metrics_analytics',
] as const;
export const REQUESTS = ['create', 'get', 'update', 'delete'] as const;

export const GLOBAL_BASIC_COUNTER: CounterType = (() => {
  const counter = {} as CounterType;
  COMPONENTS.forEach((component) => {
    counter[component] = {} as CounterType[ComponentType];
    REQUESTS.forEach((request) => {
      counter[component][request] = {
        total: 0,
      };
    });
  });
  return counter;
})();

export const DEFAULT_ROLLING_COUNTER: CounterType = (() => {
  const counter = {} as CounterType;
  COMPONENTS.forEach((component) => {
    counter[component] = {} as CounterType[ComponentType];
    REQUESTS.forEach((request) => {
      counter[component][request] = {
        count: 0,
        system_error: 0,
        user_error: 0,
      };
    });
  });
  return counter;
})();
