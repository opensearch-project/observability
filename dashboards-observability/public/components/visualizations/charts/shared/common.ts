/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { isEmpty, forEach } from 'lodash';
import { CUSTOM_LABEL } from '../../../../../common/constants/explorer';
import { ConfigList, DimensionSpan } from '../../../../../common/types/explorer';
import { removeBacktick } from '../../../../../common/utils';

export const getCompleteTimespanKey = (span: DimensionSpan) => {
  if (isEmpty(span) || isEmpty(span.time_field) || isEmpty(span.interval) || isEmpty(span.unit))
    return '';
  const spanName = `span(${span.time_field[0]?.name},${span.interval}${span.unit[0]?.value})`;
  return { name: spanName, label: spanName };
};

/**
 * Transform to traces that can be consumed by plotly.
 * @param intermediateVisData preprocessed json data that has dimensions to single aggregation mapping.
 * @param param1 required visualization configurations.
 * @returns traces.
 */
export const transformPreprocessedDataToTraces = (
  intermediateVisData: Array<any>,
  { breakdowns, isVertical = true }: Partial<ConfigList>
) => {
  const traceMap = new Map<string, any>();
  const hasBreakdown = !isEmpty(breakdowns);
  forEach(intermediateVisData, (entry) => {
    const traceKey = hasBreakdown ? [entry.breakdown, entry.aggName].join(',') : entry.aggName;
    if (isEmpty(traceMap.get(traceKey))) {
      traceMap.set(traceKey, {
        x: isVertical ? [entry.x] : [entry.value],
        y: isVertical ? [entry.value] : [entry.x],
        name: hasBreakdown ? [entry.breakdown, entry.aggName].join(',') : `${traceKey}`,
      });
    } else {
      const curTrace = traceMap.get(traceKey);
      const xaxisValue = isVertical ? entry.x : entry.value;
      const yaxisValue = isVertical ? entry.value : entry.x;
      curTrace!.x.push(xaxisValue);
      curTrace!.y.push(yaxisValue);
    }
  });
  return [...traceMap.values()];
};

/**
 * preprocess json data to
 * 1. concatenate dimensions to generate one dimension
 * 2. concatenate breakdowns (if there's any) generate one breakdown
 * 3. map dimension/breakdown to aggregations
 * @param visJson raw json data from data fetching
 * @param param1 required visualization configurations.
 * @returns intermediate visualization json data
 */
export const preprocessJsonData = (
  visJson: Array<any>,
  { dimensions, series, breakdowns, span }: Partial<ConfigList>
) => {
  const seriesFlattenedEntries = [];
  forEach(visJson, (entry) => {
    const backtickRemovedEntry = {};
    // remove backtick, so data in jsonData can be accessed through using field name
    forEach(entry, (value, key) => {
      backtickRemovedEntry[removeBacktick(key)] = value;
    });
    forEach(series, (sr) => {
      let tabularVizData = {};
      const serieKey = sr[CUSTOM_LABEL] ? sr[CUSTOM_LABEL] : `${sr.aggregation}(${sr.name})`;
      if (!isEmpty(serieKey)) {
        const concatedXaxisLabel = [
          ...(!isEmpty(span) ? [getCompleteTimespanKey(span)] : []),
          ...dimensions,
        ]
          .map((dimension) => {
            return backtickRemovedEntry[removeBacktick(dimension.name)] ?? '';
          })
          .join(',');
        const concatedBreakdownLabel = breakdowns
          ? breakdowns
              .map((breakdown) => backtickRemovedEntry[removeBacktick(breakdown.name)])
              .join(',')
          : '';
        tabularVizData = {
          value: backtickRemovedEntry[serieKey],
          x: concatedXaxisLabel,
          breakdown: concatedBreakdownLabel,
          aggName: serieKey,
        };
      } else {
        tabularVizData = {
          value: 0,
          x: '',
          breakdown: '',
        };
      }
      seriesFlattenedEntries.push(tabularVizData);
    });
  });
  return seriesFlattenedEntries;
};
