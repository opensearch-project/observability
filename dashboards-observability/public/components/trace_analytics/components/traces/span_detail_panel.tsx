/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
/* eslint-disable react-hooks/exhaustive-deps */

import {
  EuiBadge,
  EuiButtonGroup,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHorizontalRule,
  EuiPanel,
  EuiSpacer,
} from '@elastic/eui';
import _ from 'lodash';
import React, { useEffect, useMemo, useState } from 'react';
import { HttpSetup } from '../../../../../../../src/core/public';
import { Plt } from '../../../visualizations/plotly/plot';
import { TraceAnalyticsMode } from '../../home';
import { handleSpansGanttRequest } from '../../requests/traces_request_handler';
import { PanelTitle } from '../common/helper_functions';
import { SpanDetailFlyout } from './span_detail_flyout';
import { SpanDetailTable } from './span_detail_table';

export function SpanDetailPanel(props: {
  http: HttpSetup;
  traceId: string;
  colorMap: any;
  mode: TraceAnalyticsMode
  page?: string;
  openSpanFlyout?: any;
  data?: { gantt: any[]; table: any[]; ganttMaxX: number };
  setData?: (data: { gantt: any[]; table: any[]; ganttMaxX: number }) => void;
}) {
  const { mode } = props;
  const storedFilters = sessionStorage.getItem('TraceAnalyticsSpanFilters');
  const fromApp = props.page === 'app';
  const [spanFilters, setSpanFilters] = useState<Array<{ field: string; value: any }>>(
    storedFilters ? JSON.parse(storedFilters) : []
  );
  const [DSL, setDSL] = useState<any>({});
  let data: { gantt: any[]; table: any[]; ganttMaxX: number },
    setData: (data: { gantt: any[]; table: any[]; ganttMaxX: number }) => void;
  if (props.data && props.setData) {
    [data, setData] = [props.data, props.setData];
  } else {
    [data, setData] = useState<{ gantt: any[]; table: any[]; ganttMaxX: number }>({
      gantt: [],
      table: [],
      ganttMaxX: 0,
    });
  }

  const setSpanFiltersWithStorage = (newFilters: Array<{ field: string; value: any }>) => {
    setSpanFilters(newFilters);
    sessionStorage.setItem('TraceAnalyticsSpanFilters', JSON.stringify(newFilters));
  };

  const addSpanFilter = (field: string, value: any) => {
    const newFilters = [...spanFilters];
    const index = newFilters.findIndex(({ field: filterField }) => field === filterField);
    if (index === -1) {
      newFilters.push({ field, value });
    } else {
      newFilters.splice(index, 1, { field, value });
    }
    setSpanFiltersWithStorage(newFilters);
  };

  const removeSpanFilter = (field: string) => {
    const newFilters = [...spanFilters];
    const index = newFilters.findIndex(({ field: filterField }) => field === filterField);
    if (index !== -1) {
      newFilters.splice(index, 1);
      setSpanFiltersWithStorage(newFilters);
    }
  };

  const refresh = _.debounce(() => {
    if (_.isEmpty(props.colorMap)) return;
    const refreshDSL = spanFiltersToDSL();
    setDSL(refreshDSL);
    handleSpansGanttRequest(props.traceId, props.http, setData, props.colorMap, refreshDSL, mode);
  }, 150);

  const spanFiltersToDSL = () => {
    const spanDSL: any = mode === 'jaeger' ? {
      query: {
        bool: {
          must: [
            {
              term: {
                traceID: props.traceId,
              },
            },
          ],
          filter: [],
          should: [],
          must_not: [],
        },
      },
    } : {
      query: {
        bool: {
          must: [
            {
              term: {
                traceId: props.traceId,
              },
            },
          ],
          filter: [],
          should: [],
          must_not: [],
        },
      },
    };
    spanFilters.map(({ field, value }) => {
      if (value != null) {
        spanDSL.query.bool.must.push({
          term: {
            [field]: value,
          },
        });
      }
    });
    return spanDSL;
  };

  useEffect(() => {
    refresh();
  }, [props.colorMap, spanFilters]);

  const getSpanDetailLayout = (plotTraces: Plotly.Data[], maxX: number): Partial<Plotly.Layout> => {
    // get unique labels from traces
    const yLabels = plotTraces
      .map((d) => d.y[0])
      .filter((label, i, self) => self.indexOf(label) === i);
    // remove uuid when displaying y-ticks
    const yTexts = yLabels.map((label) => label.substring(0, label.length - 36));

    return {
      height: 25 * plotTraces.length + 60,
      width: 800,
      margin: {
        l: 260,
        r: 5,
        b: 30,
        t: 30,
      },
      xaxis: {
        ticksuffix: ' ms',
        side: 'top',
        color: '#91989c',
        showline: true,
        range: [0, maxX * 1.2],
      },
      yaxis: {
        showgrid: false,
        tickvals: yLabels,
        ticktext: yTexts,
      },
    };
  };

  const layout = useMemo(
    () => getSpanDetailLayout(data.gantt, data.ganttMaxX),
    [data.gantt, data.ganttMaxX]
  );

  const [currentSpan, setCurrentSpan] = useState('');

  const onClick = (event: any) => {
    if (!event?.points) return;
    const point = event.points[0];
    if (fromApp) {
      props.openSpanFlyout(point.data.spanId);
    } else {
      setCurrentSpan(point.data.spanId);
    }
  };

  const renderFilters = useMemo(() => {
    return spanFilters.map(({ field, value }) => (
      <EuiFlexItem grow={false} key={`span-filter-badge-${field}`}>
        <EuiBadge
          iconType="cross"
          iconSide="right"
          iconOnClick={() => removeSpanFilter(field)}
          iconOnClickAriaLabel="remove current filter"
        >
          {`${field}: ${value}`}
        </EuiBadge>
      </EuiFlexItem>
    ));
  }, [spanFilters]);

  const onHover = () => {
    const dragLayer = document.getElementsByClassName('nsewdrag')?.[0];
    dragLayer.style.cursor = 'pointer';
  };

  const onUnhover = () => {
    const dragLayer = document.getElementsByClassName('nsewdrag')?.[0];
    dragLayer.style.cursor = '';
  };

  const toggleOptions = [
    {
      id: 'timeline',
      label: 'Timeline',
    },
    {
      id: 'span_list',
      label: 'Span list',
    },
  ];
  const [toggleIdSelected, setToggleIdSelected] = useState(toggleOptions[0].id);

  const spanDetailTable = useMemo(
    () => (
      <SpanDetailTable
        http={props.http}
        hiddenColumns={mode === 'jaeger' ? ['traceID', 'traceGroup'] : ['traceId', 'traceGroup']}
        DSL={DSL}
        mode={mode}
        openFlyout={(spanId: string) => {
          if (fromApp) {
            props.openSpanFlyout(spanId);
          } else {
            setCurrentSpan(spanId);
          }
        }}
      />
    ),
    [DSL, setCurrentSpan]
  );

  return (
    <>
      <EuiPanel>
        <EuiFlexGroup>
          <EuiFlexItem>
            <PanelTitle title="Spans" totalItems={data.gantt.length / 2} />
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiButtonGroup
              legend="Select view of spans"
              options={toggleOptions}
              idSelected={toggleIdSelected}
              onChange={(id) => setToggleIdSelected(id)}
            />
          </EuiFlexItem>
        </EuiFlexGroup>
        {spanFilters.length > 0 && (
          <>
            <EuiSpacer size="s" />
            <EuiFlexGroup gutterSize="s" wrap>
              {renderFilters}
            </EuiFlexGroup>
          </>
        )}
        <EuiHorizontalRule margin="m" />
        <div style={{ overflowY: 'auto', maxHeight: 500 }}>
          {toggleIdSelected === 'timeline' ? (
            <Plt
              data={data.gantt}
              layout={layout}
              onClickHandler={onClick}
              onHoverHandler={onHover}
              onUnhoverHandler={onUnhover}
            />
          ) : (
            spanDetailTable
          )}
        </div>
      </EuiPanel>
      {!!currentSpan && (
        <SpanDetailFlyout
          http={props.http}
          spanId={currentSpan}
          isFlyoutVisible={!!currentSpan}
          closeFlyout={() => setCurrentSpan('')}
          addSpanFilter={addSpanFilter}
          mode={mode}
        />
      )}
    </>
  );
}
