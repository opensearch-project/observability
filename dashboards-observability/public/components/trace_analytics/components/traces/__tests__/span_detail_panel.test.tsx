/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render } from '@testing-library/react';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { SpanDetailPanel } from '../span_detail_panel';
import { BarOrientation } from '../../../../../../common/constants/shared';

describe('Service breakdown panel component', () => {
  configure({ adapter: new Adapter() });

  it('renders service breakdown panel', () => {
    const data = {
      gantt: [
        {
          x: [0],
          y: ['inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2'],
          marker: { color: 'rgba(0, 0, 0, 0)' },
          width: 0.4,
          type: 'bar',
          orientation: BarOrientation.horizontal,
          hoverinfo: 'none',
          showlegend: false,
        },
        {
          x: [19.91],
          y: ['inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2'],
          text: ['Error'],
          textfont: { color: ['#c14125'] },
          textposition: 'outside',
          marker: { color: '#7492e7' },
          width: 0.4,
          type: 'bar',
          orientation: BarOrientation.horizontal,
          hovertemplate: '%{x}<extra></extra>',
        },
      ] as Plotly.Data[],
      table: [
        {
          service_name: 'inventory',
          span_id: '32c641131b569afa',
          latency: 19.91,
          vs_benchmark: 0,
          error: 'Error',
          start_time: '2020-11-10T17:55:45.219652629Z',
          end_time: '2020-11-10T17:55:45.239564396Z',
        },
      ],
      ganttMaxX: 19.91,
    };
    const wrapper = mount(<SpanDetailPanel data={data} setData={jest.fn()} mode='data_prepper' />);
    expect(wrapper).toMatchSnapshot();
  });
});
