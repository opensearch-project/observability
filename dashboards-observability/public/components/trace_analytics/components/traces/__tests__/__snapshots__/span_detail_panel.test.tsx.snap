// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service breakdown panel component renders service breakdown panel 1`] = `
<SpanDetailPanel
  data={
    Object {
      "gantt": Array [
        Object {
          "hoverinfo": "none",
          "marker": Object {
            "color": "rgba(0, 0, 0, 0)",
          },
          "orientation": "h",
          "showlegend": false,
          "type": "bar",
          "width": 0.4,
          "x": Array [
            0,
          ],
          "y": Array [
            "inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2",
          ],
        },
        Object {
          "hovertemplate": "%{x}<extra></extra>",
          "marker": Object {
            "color": "#7492e7",
          },
          "orientation": "h",
          "text": Array [
            "Error",
          ],
          "textfont": Object {
            "color": Array [
              "#c14125",
            ],
          },
          "textposition": "outside",
          "type": "bar",
          "width": 0.4,
          "x": Array [
            19.91,
          ],
          "y": Array [
            "inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2",
          ],
        },
      ],
      "ganttMaxX": 19.91,
      "table": Array [
        Object {
          "end_time": "2020-11-10T17:55:45.239564396Z",
          "error": "Error",
          "latency": 19.91,
          "service_name": "inventory",
          "span_id": "32c641131b569afa",
          "start_time": "2020-11-10T17:55:45.219652629Z",
          "vs_benchmark": 0,
        },
      ],
    }
  }
  mode="data_prepper"
  setData={[MockFunction]}
>
  <EuiPanel>
    <div
      className="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
    >
      <EuiFlexGroup>
        <div
          className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <EuiFlexItem>
            <div
              className="euiFlexItem"
            >
              <PanelTitle
                title="Spans"
                totalItems={1}
              >
                <EuiText
                  size="m"
                >
                  <div
                    className="euiText euiText--medium"
                  >
                    <span
                      className="panel-title"
                    >
                      Spans
                    </span>
                    <span
                      className="panel-title-count"
                    >
                       (1)
                    </span>
                  </div>
                </EuiText>
              </PanelTitle>
            </div>
          </EuiFlexItem>
          <EuiFlexItem
            grow={false}
          >
            <div
              className="euiFlexItem euiFlexItem--flexGrowZero"
            >
              <EuiButtonGroup
                idSelected="timeline"
                legend="Select view of spans"
                onChange={[Function]}
                options={
                  Array [
                    Object {
                      "id": "timeline",
                      "label": "Timeline",
                    },
                    Object {
                      "id": "span_list",
                      "label": "Span list",
                    },
                  ]
                }
              >
                <fieldset
                  className="euiButtonGroup euiButtonGroup--small euiButtonGroup--text"
                  disabled={false}
                >
                  <EuiScreenReaderOnly>
                    <legend
                      className="euiScreenReaderOnly"
                    >
                      Select view of spans
                    </legend>
                  </EuiScreenReaderOnly>
                  <div
                    className="euiButtonGroup__buttons"
                  >
                    <EuiButtonGroupButton
                      color="text"
                      element="label"
                      id="timeline"
                      isDisabled={false}
                      isIconOnly={false}
                      isSelected={true}
                      key="0"
                      label="Timeline"
                      name="random_html_id"
                      onChange={[Function]}
                      size="s"
                    >
                      <EuiButtonDisplay
                        baseClassName="euiButtonGroupButton"
                        className="euiButtonGroupButton-isSelected"
                        color="text"
                        element="label"
                        fill={true}
                        htmlFor="random_html_id"
                        isDisabled={false}
                        onClick={[Function]}
                        size="s"
                        textProps={
                          Object {
                            "className": "euiButtonGroupButton__textShift",
                            "data-text": "Timeline",
                            "ref": [Function],
                            "title": "Timeline",
                          }
                        }
                      >
                        <label
                          className="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small euiButtonGroupButton--fill euiButtonGroupButton-isSelected"
                          disabled={false}
                          htmlFor="random_html_id"
                          onClick={[Function]}
                          style={
                            Object {
                              "minWidth": undefined,
                            }
                          }
                        >
                          <EuiButtonContent
                            className="euiButton__content"
                            iconSide="left"
                            textProps={
                              Object {
                                "className": "euiButton__text euiButtonGroupButton__textShift",
                                "data-text": "Timeline",
                                "ref": [Function],
                                "title": "Timeline",
                              }
                            }
                          >
                            <span
                              className="euiButtonContent euiButton__content"
                            >
                              <span
                                className="euiButton__text euiButtonGroupButton__textShift"
                                data-text="Timeline"
                                title="Timeline"
                              >
                                <input
                                  checked={true}
                                  className="euiScreenReaderOnly"
                                  data-test-subj="timeline"
                                  disabled={false}
                                  id="random_html_id"
                                  name="random_html_id"
                                  onChange={[Function]}
                                  type="radio"
                                />
                                Timeline
                              </span>
                            </span>
                          </EuiButtonContent>
                        </label>
                      </EuiButtonDisplay>
                    </EuiButtonGroupButton>
                    <EuiButtonGroupButton
                      color="text"
                      element="label"
                      id="span_list"
                      isDisabled={false}
                      isIconOnly={false}
                      isSelected={false}
                      key="1"
                      label="Span list"
                      name="random_html_id"
                      onChange={[Function]}
                      size="s"
                    >
                      <EuiButtonDisplay
                        baseClassName="euiButtonGroupButton"
                        className=""
                        color="text"
                        element="label"
                        fill={false}
                        htmlFor="random_html_id"
                        isDisabled={false}
                        onClick={[Function]}
                        size="s"
                        textProps={
                          Object {
                            "className": "euiButtonGroupButton__textShift",
                            "data-text": "Span list",
                            "ref": [Function],
                            "title": "Span list",
                          }
                        }
                      >
                        <label
                          className="euiButtonGroupButton euiButtonGroupButton--text euiButtonGroupButton--small"
                          disabled={false}
                          htmlFor="random_html_id"
                          onClick={[Function]}
                          style={
                            Object {
                              "minWidth": undefined,
                            }
                          }
                        >
                          <EuiButtonContent
                            className="euiButton__content"
                            iconSide="left"
                            textProps={
                              Object {
                                "className": "euiButton__text euiButtonGroupButton__textShift",
                                "data-text": "Span list",
                                "ref": [Function],
                                "title": "Span list",
                              }
                            }
                          >
                            <span
                              className="euiButtonContent euiButton__content"
                            >
                              <span
                                className="euiButton__text euiButtonGroupButton__textShift"
                                data-text="Span list"
                                title="Span list"
                              >
                                <input
                                  checked={false}
                                  className="euiScreenReaderOnly"
                                  data-test-subj="span_list"
                                  disabled={false}
                                  id="random_html_id"
                                  name="random_html_id"
                                  onChange={[Function]}
                                  type="radio"
                                />
                                Span list
                              </span>
                            </span>
                          </EuiButtonContent>
                        </label>
                      </EuiButtonDisplay>
                    </EuiButtonGroupButton>
                  </div>
                </fieldset>
              </EuiButtonGroup>
            </div>
          </EuiFlexItem>
        </div>
      </EuiFlexGroup>
      <EuiHorizontalRule
        margin="m"
      >
        <hr
          className="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginMedium"
        />
      </EuiHorizontalRule>
      <div
        style={
          Object {
            "maxHeight": 500,
            "overflowY": "auto",
          }
        }
      >
        <Plt
          data={
            Array [
              Object {
                "hoverinfo": "none",
                "marker": Object {
                  "color": "rgba(0, 0, 0, 0)",
                },
                "orientation": "h",
                "showlegend": false,
                "type": "bar",
                "width": 0.4,
                "x": Array [
                  0,
                ],
                "y": Array [
                  "inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2",
                ],
              },
              Object {
                "hovertemplate": "%{x}<extra></extra>",
                "marker": Object {
                  "color": "#7492e7",
                },
                "orientation": "h",
                "text": Array [
                  "Error",
                ],
                "textfont": Object {
                  "color": Array [
                    "#c14125",
                  ],
                },
                "textposition": "outside",
                "type": "bar",
                "width": 0.4,
                "x": Array [
                  19.91,
                ],
                "y": Array [
                  "inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2",
                ],
              },
            ]
          }
          layout={
            Object {
              "height": 110,
              "margin": Object {
                "b": 30,
                "l": 260,
                "r": 5,
                "t": 30,
              },
              "width": 800,
              "xaxis": Object {
                "color": "#91989c",
                "range": Array [
                  0,
                  23.892,
                ],
                "showline": true,
                "side": "top",
                "ticksuffix": " ms",
              },
              "yaxis": Object {
                "showgrid": false,
                "ticktext": Array [
                  "inventory <br>HTTP GET ",
                ],
                "tickvals": Array [
                  "inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2",
                ],
              },
            }
          }
          onClickHandler={[Function]}
          onHoverHandler={[Function]}
          onUnhoverHandler={[Function]}
        >
          <PlotlyComponent
            config={
              Object {
                "displayModeBar": false,
              }
            }
            data={
              Array [
                Object {
                  "hoverinfo": "none",
                  "marker": Object {
                    "color": "rgba(0, 0, 0, 0)",
                  },
                  "orientation": "h",
                  "showlegend": false,
                  "type": "bar",
                  "width": 0.4,
                  "x": Array [
                    0,
                  ],
                  "y": Array [
                    "inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2",
                  ],
                },
                Object {
                  "hovertemplate": "%{x}<extra></extra>",
                  "marker": Object {
                    "color": "#7492e7",
                  },
                  "orientation": "h",
                  "text": Array [
                    "Error",
                  ],
                  "textfont": Object {
                    "color": Array [
                      "#c14125",
                    ],
                  },
                  "textposition": "outside",
                  "type": "bar",
                  "width": 0.4,
                  "x": Array [
                    19.91,
                  ],
                  "y": Array [
                    "inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2",
                  ],
                },
              ]
            }
            debug={false}
            divId="explorerPlotComponent"
            layout={
              Object {
                "autosize": true,
                "barmode": "stack",
                "height": 110,
                "hovermode": "closest",
                "legend": Object {
                  "orientation": "h",
                  "traceorder": "normal",
                },
                "margin": Object {
                  "b": 30,
                  "l": 260,
                  "r": 5,
                  "t": 30,
                },
                "showlegend": false,
                "width": 800,
                "xaxis": Object {
                  "color": "#91989c",
                  "range": Array [
                    0,
                    23.892,
                  ],
                  "showline": true,
                  "side": "top",
                  "ticksuffix": " ms",
                },
                "yaxis": Object {
                  "showgrid": false,
                  "ticktext": Array [
                    "inventory <br>HTTP GET ",
                  ],
                  "tickvals": Array [
                    "inventory <br>HTTP GET 4dec6080-61af-11eb-aee3-ef2f84ffa4a2",
                  ],
                },
              }
            }
            onClick={[Function]}
            onHover={[Function]}
            onUnhover={[Function]}
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
            useResizeHandler={true}
          >
            <div
              id="explorerPlotComponent"
              style={
                Object {
                  "height": "100%",
                  "width": "100%",
                }
              }
            />
          </PlotlyComponent>
        </Plt>
      </div>
    </div>
  </EuiPanel>
</SpanDetailPanel>
`;
